crowbar_framework/lib/remote_node.rb

```diff
@@ -18,43 +18,87 @@
 require "socket"
 
 module RemoteNode
+  extend self
 
   # Wait for the host to come up and reach a certain runlevel
   # so we can send ssh commands to it
   # @param host [string] hostname or IP
   # @param timeout [integer] timeout in seconds
 
-  def self.ready?(host, timeout = 60, sleep_time = 10, options = {})
+  def ready?(host, timeout = 60, sleep_time = 10, options = {})
     timeout_at = Time.now + timeout
-    ip         = self.resolve_host(host)
+    ip = resolve_host(host)
     msg = "Waiting for host #{host} (#{ip}); next attempt in #{sleep_time} seconds until #{timeout_at}"
 
     nobj = NodeObject.find_node_by_name(host)
     reboot_requesttime = nobj[:crowbar_wall][:wait_for_reboot_requesttime] || 0
 
-    self.wait_until(msg, timeout_at.to_i, sleep_time, options) do
-      self.port_open?(ip, 22) && self.reboot_done?(ip, reboot_requesttime, options) && self.ssh_cmd(ip, runlevel_check_cmd(host))
+    wait_until(msg, timeout_at.to_i, sleep_time, options) do
+      port_open?(ip, 22) && reboot_done?(ip, reboot_requesttime, options) && ssh_cmd(ip, runlevel_check_cmd(host))
     end
   end
 
   # wait until no other chef-clients are currently running on the host
-  def self.chef_ready?(host, timeout = 60, sleep_time = 10, options = {})
+  def chef_ready?(host, timeout = 60, sleep_time = 10, options = {})
     timeout_at = Time.now + timeout
-    ip         = self.resolve_host(host)
+    ip = resolve_host(host)
     msg = "Waiting for already running chef-client on #{host} (#{ip}); next attempt in #{sleep_time} seconds until #{timeout_at}"
-    self.wait_until(msg, timeout_at.to_i, sleep_time, options) do
-      self.port_open?(ip, 22) && !self.chef_clients_running?(ip)
+    wait_until(msg, timeout_at.to_i, sleep_time, options) do
+      port_open?(ip, 22) && !chef_clients_running?(ip)
     end
   end
 
+  def reboot_done?(host_or_ip, reboot_requesttime, options = {})
+    # if reboot_requesttime is zero, we don't know when the reboot was requested.
+    # in this case return true so the process can continue
+    if reboot_requesttime > 0
+      boottime = ssh_cmd_get_boottime(host_or_ip, options)
+      if boottime > 0 && boottime <= reboot_requesttime
+        false
+      else
+        true
+      end
+    else
+      true
+    end
+  end
+
+  # get boot time of the given host or 0 if command can not be executed
+  def ssh_cmd_get_boottime(host_or_ip, options = {})
+    logger = options.fetch(:logger, nil)
+
+    ssh_cmd = ssh_cmd_base(host_or_ip)
+    ssh_cmd << "'echo $(($(date +%s) - $(cat /proc/uptime|cut -d \" \" -f 1|cut -d \".\" -f 1)))'"
```

Comment:
Above comment applies here too.  I'm more than happy to see `<<` changed but not in this PR.  Also since '<<` vs. `push` is a purely stylistic choice, I recommend you first send a PR to https://github.com/SUSE/style-guides, otherwise there's not much point changing it in a single file.